IS sabowitz

lepidopterane-atsmith opened the pull request:
I think we’re good to go with InterestSet…

lepidopterane-atsmith committed:
IS: Sarah's Stuff, V1.0

Warning: This is unfinished, so sorry if it’s a little messy. New
Classes: Bookmark and InterestSet. Mostly client stuff but some server
things.

lepidopterane-atsmith committed:
IS V1.1

Another nonfinal version but with more of a connection between server
and client.

lepidopterane-atsmith committed:
IS: Sarah, V1.2

Not sure why my bookmark feature still isn’t working

lepidopterane-atsmith committed:
IS: Sarah's Stuff, 1.3

Control for updating InterestSets rerouted through Controller. Now, if
only I could get that memory working..

lepidopterane-atsmith committed:
IS: Sarah's Stuff, V1.4

m-check works. This will eventually be reimplemented as a JUnit test in
future versions.

lepidopterane-atsmith committed:
IS: Sarah's Stuff, V1.5

Now with stable functions to add and delete interests.

lepidopterane-atsmith committed:
InterestSystem V1.7

Almost there! Introducing the Mergesort class, which I’ll probably
rename to Datesort.

lepidopterane-atsmith committed:
Interest System V1.8

Still working out this sort of problem..

lepidopterane-atsmith committed:
Interest System V1.9

I think we’re good to go…

aslo commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -15,8 +15,9 @@
+import java.util.HashSet;
 import java.util.Map;
-import java.util.Scanner;
+import java.util.Scanner; // comment this line out for testing later
```
This comment looks out of date - you should delete it.

lepidopterane-atsmith committed:
Interest System V2.0

All client-end interactions with InterestSet require an Uuid instead of
User. This should get rid of sporadic errors.

lepidopterane-atsmith committed:
Merge branch 'master' into IS_sabowitz

aslo commented on src/codeu/chat/server/View.java:
```diff
@@ -71,7 +70,17 @@ public View(Model model) {
-
+  
+  @Override
+  public ServerInfo getInfo(){
```
You should delete this `getInfo` method since it conflicts with the `getInfo` method at the bottom of this class that was brought in when you merged master in.

aslo commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -402,15 +453,28 @@ public void invoke(List<String> args) {
+      // Find the user with the given UUID and return their username.
+      // If no user has the given UUID this will return "Anonymous".
+      private String findUsername(Uuid author) {
+        for (final UserContext user : context.allUsers()) {
```
This method shares a lot of code with `findUser` below. It would be great do reduce some of the duplication here by just calling `findUser` and returning "Anonymous" if `findUser` returns null.

aslo replied:
Update: I didn't realize that `findUser` is searching by name and `findUsername` is searching by Uuid.  For now, I'd be OK with leaving them the way they are.

aslo commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -430,12 +494,21 @@ public void invoke(List<String> args) {
       }
+      
+      private User findUser(Uuid author) {
+        for (final UserContext user : context.allUsers()) {
```
Something to keep in mind for a future optimization (but not necessary for this PR): This runs in O(n) time, but if a different data structure was used to store users, it could run in O(1).

aslo commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -228,7 +243,26 @@ private UserContext findUser(String name) {
+    //
+    // Get some infomation from server; it should be version info currently
+    //
+    panel.register("info", new Panel.Command() {
```
This block of code now appears in this class twice - when you merged in master, other changes that implemented the "info" command were added (see lines 152-165 in this class). You should delete this block of code.

aslo commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -430,12 +494,21 @@ public void invoke(List<String> args) {
         
           if (message.length() > 0) {
-            conversation.add(message);
+          	MessageContext m = conversation.add(message);
```
It doesn't look like this `m` variable is actually used, so you should revert this line to how it was before.

aslo commented on src/codeu/chat/common/InterestSet.java:
```diff
@@ -0,0 +1,96 @@
+   public HashSet<Bookmark> bookmarks; 
+   
+   // custom sizing variables for users and bookmarks
+   public int u;
```
I don't think you need to bother setting these - it looks like a premature optimization to me. I suggest deleting them, and just initializing `users` and `bookmarks` with `new HashSet<>()`, which sets an arbitrary initial size and will be dynamically resized as needed.

aslo commented on src/codeu/chat/common/InterestSet.java:
```diff
@@ -0,0 +1,96 @@
+   }
+   
+   // this constructor allows custom sizings for the sets	
+   public InterestSet(int u, int b){
```
You can delete this constructor as well.

aslo commented on src/codeu/chat/common/InterestSet.java:
```diff
@@ -0,0 +1,96 @@
+     bookmarks.add(new Bookmark(c));
+   }
+   
+  public String toString() {
```
Add the `@Override` annotation on top of this method.

aslo commented on src/codeu/chat/server/Model.java:
```diff
@@ -67,10 +69,14 @@ public int compare(Time a, Time b) {
   private final Store<Time, Message> messageByTime = new Store<>(TIME_COMPARE);
   private final Store<String, Message> messageByText = new Store<>(STRING_COMPARE);
 
+  private HashMap<Uuid, InterestSet> interestMap = new HashMap<Uuid, InterestSet>(100);
```
You can omit the `100` initial size here - looks like a premature optimization to me.

aslo commented on src/codeu/chat/server/Model.java:
```diff
@@ -67,10 +69,14 @@ public int compare(Time a, Time b) {
     userByTime.insert(user.creation, user);
     userByText.insert(user.name, user);
+    interestMap.put(user.id, new InterestSet());
+    System.out.println("NEW SIZE: "+interestMap.size());
```
There are a lot of print statements in this class. If you want to keep these print statements in here, it's a good idea to convert them to use the logger that's already part of the application (see `codeu.chat.util.Logger`). You should only do that if you think these messages will be useful to other developers. Otherwise, it might be a good idea to delete them.

aslo commented on src/codeu/chat/server/Model.java:
```diff
@@ -125,4 +131,17 @@ public void add(Message message) {
+  }
+  public void updateInterests(Uuid id, InterestSet intSet){
+    System.out.println("BEFORE: "+interestMap.size());
+    interestMap.remove(id);
```
You don't need to call remove here. `interestMap.put` alone will replace whatever was in that object's position in the map.

aslo commented on src/codeu/chat/util/Mergesort.java:
```diff
@@ -0,0 +1,55 @@
+  
+  private void orderBookmarks(){
+    for(int i = 0; i< times.size(); i++){
+      orderBookmarksHelper(times.get(i));
```
This method runs in O(n^2), because the `indexOf` method used in `orderBookmarksHelper` has to scan through the entire list to find the element it's searching for. 

Not necessary to optimize in this PR, but in the future, I'd suggest not reinventing the wheel :smile: , and just using `Collections.sort` to sort your list of Bookmarks by `Time`. You can do that by implementing a custom `Comparator`, not unlike what is done currently for other collections in [`Model.java`](https://github.com/vgomezASC/CodeU-Summer-2017/pull/11/files#diff-4c08b2e8ddca5684cd55d6070c2db1a9R49).

aslo commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -457,4 +530,371 @@ public void invoke(List<String> args) {
+        boolean isCopy = false;
+        for(Bookmark b : interests.bookmarks){
+          if(!isCopy){
+            if(b.conversation.title.equals(c.conversation.title)){
```
It might be better to compare by `conversation.id` here, since they also uniquely identify conversations.

aslo commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -457,4 +530,371 @@ public void invoke(List<String> args) {
+                           message != null;
+                           message = message.next()) {
+              System.out.println();
+              System.out.format("USER : %s\n", findUsername(message.message.author));
```
This code that displays a list of bookmarks is repeated. Can you refactor it out into a separate method, perhaps called something like `displayBookmarks(List<Bookmark> bookmarks)` so that you only have to write this logic once?

aslo commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -457,4 +530,371 @@ public void invoke(List<String> args) {
+    unsorted = new ArrayList<Time>();
+	System.out.println("updates:");
+		
+    for (Bookmark b : interests.bookmarks){
```
It would make your code cleaner to have some sense of why you're doing all this logic in here. It looks to me that it's all to make sure the user isn't shown the same updates more than once. I'd recommend either a) adding a brief comment saying that, or b) moving this code into a separate method with a descriptive name.

lepidopterane-atsmith committed:
Interest System V2.0.1

Just the beginning of some changes in Chat.java.

lepidopterane-atsmith committed:
Interest System V2.0.2

Addressed issues brought up in code review. Sorry if it says I made 72
file changes bc I changed my local codebase after solving merge
conflicts.

lepidopterane-atsmith committed:
Interest-System-V2.0.5

lepidopterane-atsmith committed:
IS_V2.0.6

lepidopterane-atsmith committed:
IS_V2.0.7

aslo commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -34,13 +34,15 @@
 import codeu.chat.common.InterestSet;
```
It looks like you've inadvertently changed the file permissions of many files in this pull request. Notice that at the top, many files display a change like this:

![](https://cloud.githubusercontent.com/assets/1039411/26559668/03f9fde4-447f-11e7-8cfa-8e09f8edb2af.png)

That means these files have been changed to an "executable" status, which is incorrect - they are not executable files, and this can be confusing to other developers in the future. More importantly, a pull request should only change the files relevant to the feature being worked on. Before merging this, I strongly recommend that you change these files back to the default permissions. Assuming you're on Linux or Mac, for each of these files, you should run chmod 0644 <file_name_here> to revert them back to the correct status.

This stackoverflow answer has more info about file permissions in case you're interested.

lepidopterane-atsmith replied:
Link?

aslo replied:
:+1:  https://stackoverflow.com/questions/737673/how-to-read-the-mode-field-of-git-ls-trees-output/8347325#8347325

aslo commented on src/codeu/chat/server/Server.java:
```diff
@@ -192,6 +192,7 @@ public void onMessage(InputStream in,OutputStream out) throws IOException
         Uuid.SERIALIZER.write(out, view.getInfo().version);
       }
     });
+<<<<<<< HEAD
```
Don't forget to remove these markers of merge conflicts in this file.

aslo reviewed (changes requested):
Getting closer! A few more things:

* You should avoid committing your `persistent-dir`, as this directory is only relevant to your local workspace and not to the other developers of this project. You should remove that directory from git by running `git rm persistent-dir/*`
* Make sure you compile the code and run some sanity checks before pushing new changes. As-is, this code doesn't compile.

lepidopterane-atsmith committed:
Interest System V2.0.8

This should have less executable files and actually work now.

lepidopterane-atsmith committed:
Interest System V2.0.9

Fixed permissions further.
